src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1323 @@
+ * running the background tasks queue monitor.
+ *
+ * It's mainloop reads a runnable task from pg_dist_background_task and progressing the
+ * tasks and jobs state machines associated with the task. When no new task can be found
+ * it will exit(0) and lets the maintenance daemon poll for new tasks.
+ *
+ * The main loop is currently implemented as a synchronous loop stepping through the task
+ * and update its state before going to the next.
+ */
+void
+CitusBackgroundTaskQueueMonitorMain(Datum arg)
+{
+	Oid databaseOid = DatumGetObjectId(arg);
+
+	/* extension owner is passed via bgw_extra */
+	Oid extensionOwner = InvalidOid;
+	memcpy_s(&extensionOwner, sizeof(extensionOwner),
+			 MyBgworkerEntry->bgw_extra, sizeof(Oid));
+
+	BackgroundWorkerUnblockSignals();
+
+	/* connect to database, after that we can actually access catalogs */
+	BackgroundWorkerInitializeConnectionByOid(databaseOid, extensionOwner, 0);
+
+	/*
+	 * There should be exactly one background task monitor running, running multiple would
+	 * cause conflicts on processing the tasks in the catalog table as well as violate
+	 * parallelism guarantees. To make sure there is at most, exactly one backend running
+	 * we take a session lock on the CITUS_BACKGROUND_TASK_MONITOR operation.
+	 *
+	 * TODO now that we have a lock, we should install a term handler to terminate any
+	 * 'child' backend when we are terminated. Otherwise we will still have a situation
+	 * where the actual task could be running multiple times.
+	 */
+	LOCKTAG tag = { 0 };
+	SET_LOCKTAG_CITUS_OPERATION(tag, CITUS_BACKGROUND_TASK_MONITOR);
+	const bool sessionLock = true;
+	const bool dontWait = true;
+	LockAcquireResult locked =
+		LockAcquire(&tag, AccessExclusiveLock, sessionLock, dontWait);
+	if (locked == LOCKACQUIRE_NOT_AVAIL)
+	{
+		ereport(ERROR, (errmsg("background task queue monitor already running for "
+							   "database")));
+	}
+
+	/* make worker recognizable in pg_stat_activity */
+	pgstat_report_appname("citus background task queue monitor");
+
+	ereport(DEBUG1, (errmsg("started citus background task queue monitor")));
+
+	/* TODO this is here for debugging purposses, remove before merge. */
+	if (BackgroundTaskMonitorDebugDelay)
+	{
+		pg_usleep(30 * 1000 * 1000);
+	}
+
+	MemoryContext perTaskContext = AllocSetContextCreate(CurrentMemoryContext,
+														 "PerTaskContext",
+														 ALLOCSET_DEFAULT_MINSIZE,
+														 ALLOCSET_DEFAULT_INITSIZE,
+														 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/*
+	 * First we find all jobs that are running, we need to check if they are still running
+	 * if not reset their state back to scheduled.
+	 */
+	{
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		ResetRunningBackgroundTasks();
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+	}
+
+
+	MemoryContext oldContextPerJob = MemoryContextSwitchTo(perTaskContext);
+	TimestampTz backgroundWorkerFailedStartTime = 0;
+	bool hasTasks = true;
```

Comment:
minor: this boolean is not really used, we break out of the loop immediately after setting it to false

Reply:
I have added a comment on why we keep an unused variable here for future reference.